extensions/amp-base-carousel/1.0/scroller.js

```diff
@@ -0,0 +1,253 @@
+  };
+
+  const handleScroll = () => {
+    if (ignoreProgrammaticScroll.current) {
+      ignoreProgrammaticScroll.current = false;
+      return;
+    }
+    updateCurrentIndex();
+    debouncedResetScrollReferencePoint();
+  };
+
+  return (
+    <>
+      <style>{styles.hideScrollbarPseudo}</style>
+      <div
+        hide-scrollbar
+        key="container"
+        ref={containerRef}
+        onScroll={handleScroll}
+        style={{
+          ...styles.scrollContainer,
+          ...styles.hideScrollbar,
+          ...styles.horizontalScroll,
+        }}
+        tabindex={0}
+      >
+        {slides}
+      </div>
+    </>
+  );
+}
+
+/**
+ * How the slides are ordered when looping:
+ *
+ * We want to make sure that the user can scroll all the way to the opposite
+ * end (either forwards or backwards) of the carousel, but no further (no
+ * looping back past where you started). We render elements dynamically
+ * for a desirable scrolling order to allow the browser to scroll as well as
+ * providing targets for the browser to snap on. This is important as these
+ * targets need to be present before the scroll starts for things to work
+ * correctly.
+ *
+ * The elements are ordered depending on the reference point, called
+ * the restingIndex to allow full movement forwards and backwards. You can
+ * imagine the DOM structure looks like the following if you have 5 slides:
+ *
+ * [1][2][3][4][5]
+ *
+ * When the restingIndex is the first index, we should translate slides as follows:
+ *
+ * [4][5][1][2][3]
+ *
+ * This ensures that if you move left or right, there is a slide to show.
+ *
+ * When the user stops scrolling, we update the restingIndex and show/hide the
+ * appropriate spacers. For example, if the user started at slide '1' and moved
+ * left to slide '4', the UI would eventually stop because there is
+ * nothing more to the left of slide '4'.
+ *
+ * [4][5][1][2][3]
+ *
+ * Once scrolling stops, however, the reference point would be reset and this would
+ * update to the following with the next render:
+ *
+ * [2][3][4][5][1]
+ *
+ * Ordering slides:
+ *
+ * Slides are ordered to be before or after the current slide and do not rearrange
+ * dynamically as the user scrolls. Only once the scrolling has ended do they rearrange.
+ * Currently, half the slides are ordered before and half the slides are ordered after.
+ * This could be a bit smarter and only place as many as are necessary on either side of
+ * the reference point to have a sufficient amount of buffer.
+ *
+ * Initial index:
+ *
+ * The initial index can be specified, which will make the carousel scroll to
+ * the desired index when it first renders.
+ *
+ * @param {!JsonObject} props
+ * @return {PreactDef.Renderable}
+ */
+function renderSlides(props) {
+  const {
+    'children': children,
+    'restingIndex': restingIndex,
+    'offsetRef': offsetRef,
+    'pivotIndex': pivotIndex,
+    'loop': loop,
+  } = props;
+  const {length} = children;
```

Comment:
Shouldn't the be on the `toChildArary` return value?

Reply:
Good point, is it OK to assume that `children` is an array when passed in as a prop to `Scroller` if we make sure the result of `toChildArray` is what we give `Scroller` in `BaseCarousel`? This will be easier to assert when we have the component prop externs.

Reply:
Now that we're passing in the return value from `toChildArray`, I think it's fine to just treat this as always being an array.